issue_674: how to debug rules in AdGuard for Chrome MV3 (#700)
AdguardTeam/KnowledgeBase: docs/adguard-browser-extension/solving-problems/debug-rules.md

    ```

1. **Extract text filters:**
    ```shell
    pnpm debug-filters extract
    ```
    - This command has `-b, --browser <browser>` option to specify the browser target.
      Available browsers: `chrome-mv3`, `opera-mv3`.
      Default: `chrome-mv3`.

1. **Edit the text filters** in `./build/dev/<browser>/filters` as needed.

1. **Convert filters to DNR rulesets:**
    ```shell
    pnpm debug-filters convert
    ```
    - This command has `-b, --browser <browser>` option to specify the browser target.
      Available browsers: `chrome-mv3`, `opera-mv3`.
      Default: `chrome-mv3`.

1. **Reload the extension in your browser** to apply new rulesets.

**Tip:**
- To download the latest available text filters, run:
    ```shell
    pnpm debug-filters load
    ```
    - This command has `-b, --browser <browser>` option to specify the browser target.
      Available browsers: `chrome-mv3`, `opera-mv3`.
      Default: `chrome-mv3`.

If you see an exclamation mark in the filtering log, it means the assumed rule (calculated by the engine) and the applied rule (converted to DNR) are different. Otherwise, only the applied rule (in DNR and text ways) will be shown.

> ##### <a name="dev-technical-info-about-debug-commands"></a> Technical information about commands

Reviewer: @105th is this section needs to be added?

## How to debug rules

For faster development of DNR rulesets, you can use `@adguard/dnr-rulesets` watch mode, which will automatically rebuild rulesets whenever filter files change:

> 1. First, build the extension once (just skip, if you did it in the "How to build extension" section):

Reviewer 2: ```suggestion
1. First, build the extension (skip this step if you already did so in the *How to build extension* section):
```

    ```shell
    pnpm dev chrome-mv3
    ```

1. Start the watch mode with all required filters:

    ```shell
    npx @adguard/dnr-rulesets watch \
        # Force update of rulesets in manifest.json
        --force-update \
        # Enable rulesets with IDs 1 and 2
        --enable=1,2 \
        # Download filters from the server on the first run
        --download \
        # Path to the extension manifest
        ./build/dev/chrome-mv3/manifest.json \
        # Path to the filters directory
        ./build/dev/chrome-mv3/filters \
        # Path to the web-accessible-resources directory
        ./Extension/web-accessible-resources \
        # Output directory for DNR rulesets
        ./build/dev/chrome-mv3/filters/declarative
    ```

    The `--download` flag will download all filters from the server on the first run. For subsequent runs, you can omit this flag to use existing filters:

    ```shell
    npx @adguard/dnr-rulesets watch \
        --force-update \
        --enable=1,2 \
        ./build/dev/chrome-mv3/manifest.json \
        ./build/dev/chrome-mv3/filters \
        ./Extension/web-accessible-resources \
        ./build/dev/chrome-mv3/filters/declarative
    ```

> 1. Now, whenever you modify filter files, the DNR rulesets will be automatically rebuilt without having to rebuild the entire extension.

Reviewer 2: ```suggestion
Now, whenever you modify filter files, the DNR rulesets will rebuild automatically, so you won’t have to rebuild the entire extension.
```

1. Reload the extension in the browser:

    ![Reload extension](https://cdn.adtidy.org/content/Kb/ad_blocker/browser_extension/reload_extension.png)

> 1. If you see an ❗ mark - it means that assumed rule (which we calculated with our tsurlfilter engine, which performed applying rules in MV2) and actually applied rule (from which we converted to DNR rule) are not the same. And this can be a problem of conversion. <br/> Otherwise, if assumed and applied rules are the same - only applied rule, in raw text and declarative rule views, will be shown.

Reviewer 2: ```suggestion
1. If you see an exclamation mark, it means the assumed rule (calculated by our tsurlfilter engine using MV2 rules) and the applied rule (converted to a DNR rule) are different. This can be a problem of conversion.

Otherwise, if assumed and applied rules are the same, only applied rule, in raw text and declarative rule views, will be shown.
```
